search/mongo_search.go

```diff
@@ -97,44 +121,59 @@ func (m *MongoSearcher) Search(query Query) (results interface{}, total uint32,
 		return nil, 0, err
 	}
 
+	// If the count wasn't already in cache, add it to cache.
+	if m.readonly && docount {
+		countcache := &CountCache{
+			Id:    queryHash,
+			Count: total,
+		}
+		// Don't collect the error here since this should fail silently.
+		m.db.C("countcache").Insert(countcache)
+	}
+
 	return results, total, nil
 }
 
 // aggregate takes a BSONQuery and runs its Pipeline through the mongo aggregation framework. Any query options
 // will be added to the end of the pipeline.
-func (m *MongoSearcher) aggregate(bsonQuery *BSONQuery, options *QueryOptions) (pipe *mgo.Pipe, total uint32, err error) {
+func (m *MongoSearcher) aggregate(bsonQuery *BSONQuery, options *QueryOptions, docount bool) (pipe *mgo.Pipe, total uint32, err error) {
 	c := m.db.C(models.PluralizeLowerResourceName(bsonQuery.Resource))
 
 	// First get a count of the total results (doesn't apply any options)
-	if len(bsonQuery.Pipeline) == 1 {
-		// The pipeline is only being used for includes/revincludes, meaning the entire
-		// collection is being searched. It's faster just to get a total count from the
-		// collection after a find operation. The first stage in the Pipeline will
-		// always be a $match stage.
-		intTotal, err := c.Find(bsonQuery.Pipeline[0]["$match"]).Count()
-		if err != nil {
-			return nil, 0, err
+	if docount {
+		if len(bsonQuery.Pipeline) == 1 {
+			// The pipeline is only being used for includes/revincludes, meaning the entire
+			// collection is being searched. It's faster just to get a total count from the
+			// collection after a find operation. The first stage in the Pipeline will
+			// always be a $match stage.
+			intTotal, err := c.Find(bsonQuery.Pipeline[0]["$match"]).Count()
+			if err != nil {
+				return nil, 0, err
+			}
+			total = uint32(intTotal)
+		} else {
+			// Do the count in the aggregation framework
+			countStage := bson.M{"$group": bson.M{
+				"_id":   nil,
+				"total": bson.M{"$sum": 1},
+			}}
+			countPipeline := make([]bson.M, len(bsonQuery.Pipeline)+1)
+			copy(countPipeline, bsonQuery.Pipeline)
+			countPipeline[len(countPipeline)-1] = countStage
+
+			result := struct {
+				Total float64 `bson:"total"`
+			}{}
+
+			err = c.Pipe(countPipeline).One(&result)
+			if err != nil {
+				return nil, 0, err
+			}
+			total = uint32(result.Total)
 		}
-		total = uint32(intTotal)
 	} else {
-		// Do the count in the aggregation framework
-		countStage := bson.M{"$group": bson.M{
-			"_id":   nil,
-			"total": bson.M{"$sum": 1},
-		}}
-		countPipeline := make([]bson.M, len(bsonQuery.Pipeline)+1)
-		copy(countPipeline, bsonQuery.Pipeline)
-		countPipeline[len(countPipeline)-1] = countStage
-
-		result := struct {
-			Total float64 `bson:"total"`
-		}{}
-
-		err = c.Pipe(countPipeline).One(&result)
-		if err != nil {
-			return nil, 0, err
-		}
-		total = uint32(result.Total)
+		// The count was already cached, so return 0 and expect it to be overwritten in MongoSearcher.Search().
+		total = uint32(0)
```

Comment:
Do you even need to do this? If you never touch total, will go just return the zero value for the type or does go make you explicitly set it?